docs/guide/migration.md

```diff
@@ -1,133 +1,48 @@
-
-```js
-import vue from '@vitejs/plugin-vue'
-import { defineConfig } from 'vite'
-
-export default defineConfig({
-  plugins: [vue()]
-})
-```
-
-### Custom Blocks Transforms
-
-A custom plugin can be used to transform Vue custom blocks like the one below:
-
-```ts
-// vite.config.js
-import vue from '@vitejs/plugin-vue'
-import { defineConfig } from 'vite'
-
-const vueI18nPlugin = {
-  name: 'vue-i18n',
-  transform(code, id) {
-    if (!/vue&type=i18n/.test(id)) {
-      return
-    }
-    if (/\.ya?ml$/.test(id)) {
-      code = JSON.stringify(require('js-yaml').load(code.trim()))
-    }
-    return `export default Comp => {
-      Comp.i18n = ${code}
-    }`
-  }
-}
-
-export default defineConfig({
-  plugins: [vue(), vueI18nPlugin]
-})
-```
+## Config Options Change
 
-## React Support
+- The following options that were already deprecated in v2 have been removed:
 
-React Fast Refresh support is now provided via [`@vitejs/plugin-react`](https://github.com/vitejs/vite/tree/main/packages/plugin-react).
+  - `optimizeDeps.keepNames` (switch to [`optimizeDeps.esbuildOptions.keepNames`](../config/dep-optimization-options.md#optimizedepsesbuildoptions))
+  - `build.base` (switch to [`base`](../config/shared-options.md#base))
+  - `alias` (switch to [`resolve.alias`](../config/shared-options.md#resolvealias))
+  - `dedupe` (switch to [`resolve.dedupe`](../config/shared-options.md#resolvededupe))
+  - `polyfillDynamicImport` (use [`@vitejs/plugin-legacy](https://github.com/vitejs/vite/tree/main/packages/plugin-legacy) for browsers without dynamic import support)
 
-## HMR API Change
+## Dev server Changes
 
-`import.meta.hot.acceptDeps()` have been deprecated. [`import.meta.hot.accept()`](./api-hmr#hot-accept-deps-cb) can now accept single or multiple deps.
+Vite optimizes dependencies with esbuild to both convert CJS only deps to ESM and to reduce the number of modules the browser needs to request. In v3 the default strategy to discover and batch dependencies has changed. Vite no longer pre-scan user code with esbuild to get an initial list of dependencies on cold start. Instead, it delays the first dependency optimization run until every imported user module on load is processed. To get back the v2 strategy you can use [`optimizeDeps.devScan`](../config/dep-optimization-options.md#optimizedepsdevscan).
 
-## Manifest Format Change
+## Build Changes
 
-The build manifest now uses the following format:
+In v3, Vite uses esbuild to optimize dependencies by default. Doing so, it removes one of the most significant differences between dev and prod present in v2. Because esbuild converts CJS only dependencies to ESM, [`@rollupjs/plugin-commonjs`] is no longer used. If you need to get back to the v2 strategy, you can use [`optimizeDeps.disabled: 'build'`](../config/dep-optimization-options.md#optimizedepsdisabled).
```

Comment:
```suggestion
In v3, Vite uses esbuild to optimize dependencies by default. Doing so, it removes one of the most significant differences between dev and prod present in v2. Because esbuild converts CJS-only dependencies to ESM, [`@rollupjs/plugin-commonjs`] is no longer used. If you need to get back to the v2 strategy, you can use [`optimizeDeps.disabled: 'build'`](../config/dep-optimization-options.md#optimizedepsdisabled).
```